### PR TITLE
Custom densities

### PR DESCRIPTION
**Pull request by bsculac**
## What is the change?

The ability to specify custom densities for non-custom materials is added in this PR.

## Why is the change being made?

Previously, only materials designated "Custom" in the blueprints file were allowed to have custom densities. This change allows perturbation of the component density as the component is generated. This is specifically done to prevent modification of the base material properties, which could have unintended downstream effects.

This PR attempts to resolve https://github.com/terrapower/armi/issues/1272. The use case is for adjustment of library materials to account for changes in density during fabrication (e.g. porosity) without having to use workarounds like smear density adjustments.

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/reactor/blueprints/isotopicOptions.py [hunk 1]
```diff
@@ -393,15 +393,11 @@ def apply(self, material):
...
-                    "(which implies a mass density) on `{}` with custom isotopics `{}`. "
-                    "This has no effect on this Material class; you can only "
-                    "override mass density on `Custom` "
-                    "materials. Consider switching to number fraction input. "
-                    "Continuing to use {} mass density.".format(
-                        material, self.name, material
-                    )
+                runLog.important(
+                    "A custom density or number densities has been specified for a non-custom "
+                    "material. The material density will not be updated to prevent unintentional "
```
Comment (onufer) [inline armi/reactor/blueprints/isotopicOptions.py:398]: `material object's density`
or
"The component's density has been updated to reflect the change, but the material it points to preserves the actual material density (just like for 2d expansion)"

File: armi/reactor/blueprints/tests/test_customIsotopics.py [hunk 2]
```diff
@@ -216,18 +220,23 @@ def test_massFractionsAreApplied(self):
...
 
-        # density only works with a Custom material type.
-        self.assertAlmostEqual(fuel0.density(), fuel2.density())
         self.assertEqual(
             set(fuel2.p.numberDensities.keys()), set(fuel1.p.numberDensities.keys())
         )  # keys are same
 
+        self.assertAlmostEqual(19.1, fuel2.density())
+        # original material density should not be changed after setting a custom density component
+        self.assertNotAlmostEqual(fuel2.density(), fuel6.density(), places=2)
```
Comment (onufer) [inline armi/reactor/blueprints/tests/test_customIsotopics.py:235]: change the fuel2 density to something besides 19.1 so that we can confirm it actually happened. these tests seem to indicate the baseline density is 19.1, so the test hasn't confirmed a change

Comment (bsculac) [reply armi/reactor/blueprints/tests/test_customIsotopics.py:235]: The density of 19.1 is what is set in the blueprints above. Fuel blocks 0, 2, and 6 are all using the non-custom material. Fuel blocks 0 and 6 do not have a custom density set, while fuel block 2 has a custom density. Line 235 verifies that the custom density of fuel block 2 is not equal to fuel block 6, therefore the density of fuel block 2 has been changed from the baseline material density.

This test originally tested that fuel block 2 did _not_ change density because the material was not set to custom. Let me know if additional clarification is needed.

Comment (onufer) [reply armi/reactor/blueprints/tests/test_customIsotopics.py:235]: ooooOOOOoo. its really easy for my brain to skip the  word "not"

could you please assert it is equal to whatever its equal to so its more black and white what the actual density is?

File: armi/reactor/blueprints/componentBlueprint.py [hunk 3]
```diff
@@ -225,8 +225,33 @@ class can then use to modify the isotopics as necessary.
...
 
+    def setCustomDensity(self, constructedComponent, blueprint):
+        """Apply a custom density to a material with custom isotopics but not a 'custom material'."""
+        if self.isotopics is not None:
+            # must nest because one can't check for attribues on a NoneType
+            if blueprint.customIsotopics[self.isotopics].density is not None:
+                mat = materials.resolveMaterialClassByName(self.material)()
+                if not isinstance(mat, materials.Custom):
+                    densityRatio = (
+                        blueprint.customIsotopics[self.isotopics].density
```
Comment (onufer) [inline armi/reactor/blueprints/componentBlueprint.py:242]: do we always have a density? i thought some custom isotopic input options maybe didnt specify density. maybe they all eventually resolve it though

**Comment by keckler** [comment]
At what temperature is the entered density supposed to correspond to?

**Comment by keckler** [comment]
This PR should probably update this section of the docs: 
https://terrapower.github.io/armi/user/inputs.html#custom-isotopics

**Comment by bsculac** [comment]
> At what temperature is the entered density supposed to correspond to?

The density is set at component construction so it should be the cold dimension unless the blueprints was specified to be at hot dimensions.

File: armi/reactor/blueprints/componentBlueprint.py (hunk 3 above)
Comment (bsculac) [reply armi/reactor/blueprints/componentBlueprint.py:242]: No, the custom isotopics entry can be specified without a density, which is why line 238 checks for it. But if `customIsotopics` exists then it's attributes are initialized to `None`

File: armi/reactor/blueprints/isotopicOptions.py (hunk 1 above)
Comment (bsculac) [reply armi/reactor/blueprints/isotopicOptions.py:398]:
```suggestion
                    "material. The material object's density will not be updated to prevent unintentional "
```

**Comment by keckler** [comment]
> > At what temperature is the entered density supposed to correspond to?
> 
> The density is set at component construction so it should be the cold dimension unless the blueprint was specified to be at hot dimensions.

I suppose I am asking for this to be outlined somewhere, for instance in the blueprints documentation.

**Comment by john-science** [comment]
I would love a unit test that actually tests the new feature you are building here. Show me the _goal_ in your unit test, and show me the edge cases.

And example edge case might be proving that this PR fixes the bug in #1272.

As always, if that's too much of a lift I can help.

File: armi/reactor/blueprints/tests/test_customIsotopics.py [hunk 4]
```diff
@@ -179,10 +192,168 @@ class TestCustomIsotopics(unittest.TestCase):
...
+
+    # >>> from armi.nucDirectory import elements, nuclideBases
+    # >>> import numpy
+    # >>> u = elements.bySymbol['U']
+    # >>> Mi = numpy.array([n.weight for n in u.getNaturalIsotopics()])
+    # >>> w_i = numpy.array([n.abundance for n in u.getNaturalIsotopics()])
+    # >>> Ni = 19.1 * w_i * 6.0221e23 / Mi
+    # array([  2.65398007e+18,   3.52549755e+20,   4.79692055e+22])
+    # >>> for n, ni in zip(u.getNaturalIsotopics(), Ni):
+    # >>>     print '        {}: {:.7e}'.format(n.name, ni) # requires 7 decimal places!
```
Comment (john-science) [inline armi/reactor/blueprints/tests/test_customIsotopics.py:279]: Is this bit an error? It looks like random Python code inside a YAML file.

File: armi/tests/tutorials/c5g7-blueprints.yaml [hunk 5]
```diff
@@ -162,7 +162,6 @@ blocks:
             # has same od as fuel, so there needs to be something in the gap.
             shape: Circle
             material: Void
-            isotopics: moderator
```
Comment (john-science) [inline armi/tests/tutorials/c5g7-blueprints.yaml:165]: Sorry, honest question, why did we need to make this change?

(Just trying to wrap my head around the PR.  I am not worried either way.)

File: armi/reactor/blueprints/componentBlueprint.py [hunk 6]
```diff
@@ -225,8 +225,52 @@ class can then use to modify the isotopics as necessary.
...
+                        / constructedComponent.density()
+                    )
+                    constructedComponent.changeNDensByFactor(densityRatio)
+
+                    runLog.important(
+                        "A custom material density was specified in the custom isotopics for non-custom "
+                        "material {} in component {}. The component density has been altered to "
+                        "{}.".format(
+                            mat, constructedComponent, constructedComponent.density()
+                        )
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:271]:
```suggestion
                        ), single=True
```

File: armi/reactor/blueprints/componentBlueprint.py [hunk 7]
```diff
@@ -225,8 +225,52 @@ class can then use to modify the isotopics as necessary.
...
+                    )
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:272]:
```suggestion
                    runLog.important(
                        "A custom material density was specified in the custom isotopics for non-custom "
                        "material {} in component {}. The component density has been altered to "
                        "{}.".format(
                            mat, constructedComponent, constructedComponent.density()
                        ),
                        single=True
                    )
```

Comment (john-science) [reply armi/reactor/blueprints/componentBlueprint.py:272]: What I am suggesting here is that we make sure this log message is only ever printed once. Imagine if this message were printed hundreds of times for a single reactor, during construction. Boring, right?  This `single=True` should make sure no logging duplication happens.

File: armi/tests/tutorials/c5g7-blueprints.yaml (hunk 5 above)
Comment (bsculac) [reply armi/tests/tutorials/c5g7-blueprints.yaml:165]: This PR does not allow density to be assigned to Void and throws an error if a user attempts to do so. Previously, this would weakly fail, emitting a warning that nothing was done.

The reason for throwing an error (aside from it being silly to assign properties to Void) was to avoid dividing by zero.

File: armi/reactor/blueprints/componentBlueprint.py (hunk 7 above)
Comment (bsculac) [reply armi/reactor/blueprints/componentBlueprint.py:272]:
```suggestion
                    runLog.important(
                        "A custom material density was specified in the custom isotopics for non-custom "
                        "material {}. The component density has been altered to "
                        "{}.".format(
                            mat, constructedComponent.density()
                        ),
                        single=True
                    )
```
Good point, edited the suggestion to remove the component print to make the messages non-unique.

File: armi/reactor/blueprints/isotopicOptions.py [hunk 8]
```diff
@@ -393,15 +393,11 @@ def apply(self, material):
...
-                    "materials. Consider switching to number fraction input. "
-                    "Continuing to use {} mass density.".format(
-                        material, self.name, material
-                    )
+                runLog.important(
+                    "A custom density or number densities has been specified for non-custom "
+                    "material {}. The material object's density will not be updated to prevent unintentional "
+                    "density changes across the model. Only custom materials may have a density "
+                    "specified.".format(material)
                 )
```
Comment (bsculac) [inline armi/reactor/blueprints/isotopicOptions.py:401]:
```suggestion
                runLog.important(
                    "A custom density or number densities has been specified for non-custom "
                    "material {}. The material object's density will not be updated to prevent unintentional "
                    "density changes across the model. Only custom materials may have a density "
                    "specified.".format(material),
                    single=True
                )
```
Incorporate `single=True` suggestion from other comment.

File: armi/reactor/blueprints/tests/test_customIsotopics.py (hunk 4 above)
Comment (john-science) [reply armi/reactor/blueprints/tests/test_customIsotopics.py]: Score, thanks!

File: armi/reactor/blueprints/componentBlueprint.py [hunk 9]
```diff
@@ -225,8 +225,51 @@ class can then use to modify the isotopics as necessary.
...
+        # set the custom density for non-custom material components after construction
+        self.setCustomDensity(constructedObject, blueprint, matMods)
+
         return constructedObject
 
+    def setCustomDensity(self, constructedComponent, blueprint, matMods):
+        """Apply a custom density to a material with custom isotopics but not a 'custom material'."""
+        if self.isotopics is not None:
+            # must nest because one can't check for attribues on a NoneType
+            if blueprint.customIsotopics[self.isotopics].density > 0:
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:238]: These two `if` blocks make the entire rest of the method double-indented. Another option would be to do:

```python
        if self.isotopics is None:
            return

    if blueprint.customIsotopics[self.isotopics].density <= 0:
            return
```

This might make the rest of the method easier to read. But I'm not going to die on any hills today, I'm just pointing it out for the future.

Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:238]: If I go into the Python interpreter and do this:

```bash
>>> None > 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

If the `.density` is `None`, this line doesn't return `False`, it raises a `TypeError`.

Comment (bsculac) [reply armi/reactor/blueprints/componentBlueprint.py:238]: This is a good suggestion but it makes me realize that a user specifying a zero or negative density is not handled. I will be pushing something up that handles said case.

Comment (bsculac) [reply armi/reactor/blueprints/componentBlueprint.py]: I've altered the logic in this section, let me know what you think.

**Review by john-science — APPROVED** [review]

File: armi/reactor/blueprints/componentBlueprint.py [hunk 10]
```diff
@@ -225,8 +225,67 @@ class can then use to modify the isotopics as necessary.
...
 
+    def setCustomDensity(self, constructedComponent, blueprint, matMods):
+        """Apply a custom density to a material with custom isotopics but not a 'custom material'."""
+        if self.isotopics is None:
+            # No custom isotopics specified
+            return
+        density = blueprint.customIsotopics[self.isotopics].density
+        if density is None:
+            # Nothing to do
+            return
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:242]:
```suggestion
            return

        density = blueprint.customIsotopics[self.isotopics].density
        if density is None:
            # Nothing to do
            return

```

File: armi/reactor/blueprints/componentBlueprint.py [hunk 11]
```diff
@@ -225,8 +225,69 @@ class can then use to modify the isotopics as necessary.
...
+                    )
+                )
+                raise ValueError(
+                    "Cannot apply custom densities to materials without density."
+                )
+
+            densityRatio = (
+                blueprint.customIsotopics[self.isotopics].density
+                / constructedComponent.density()
+            )
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:281]:
```suggestion
            densityRatio = density / constructedComponent.density()
```

File: armi/reactor/blueprints/componentBlueprint.py [hunk 12]
```diff
@@ -225,8 +225,69 @@ class can then use to modify the isotopics as necessary.
...
+        if self.isotopics is None:
+            # No custom isotopics specified
+            return
+
+        density = blueprint.customIsotopics[self.isotopics].density
+        if density is None:
+            # Nothing to do
+            return
+
+        if blueprint.customIsotopics[self.isotopics].density <= 0:
```
Comment (john-science) [inline armi/reactor/blueprints/componentBlueprint.py:245]:
```suggestion
        if density <= 0:
```

Comment (john-science) [reply armi/reactor/blueprints/componentBlueprint.py:245]: Just seems easier to read. NBD. I won't die on this hill.

File: armi/reactor/blueprints/componentBlueprint.py (hunk 11 above)
Comment (john-science) [reply armi/reactor/blueprints/componentBlueprint.py:281]: Just seems easier to read. NBD. I won't die on this hill.

File: armi/reactor/blueprints/tests/test_customIsotopics.py [hunk 13]
```diff
@@ -216,18 +332,89 @@ def test_massFractionsAreApplied(self):
...
+        # The block with custom density but not the 'Custom' material
+        fuel2 = self.a[2].getComponent(Flags.FUEL)
+        # A block like the template block, but made after the custom block
+        fuel6 = self.a[6].getComponent(Flags.FUEL)
+        # A block with custom density set via number density
+        fuel8 = self.a[8].getComponent(Flags.FUEL)
+
+        # Check that the density is set correctly on the custom density block,
+        # and that it is not the same as the original
+        self.assertAlmostEqual(19.1, fuel2.density())
```
[the remark this answers is not in the document]
Comment (bsculac) [reply armi/reactor/blueprints/tests/test_customIsotopics.py:357]: If you are asking if the custom density logic should be adjusting to the working temperature, probably. When I wrote this PR, I had misunderstood what Tin in the blueprints was used for, see https://github.com/terrapower/armi/issues/1818#issuecomment-2289766135 for more details. In short, there is an assumption of hot heights in the blueprints unless the option `inputHeightsConsideredHot` is `False`, so in the default case of `inputHeightsConsideredHot`=True the feature in this PR will assign incorrect densities for components with custom isotopics because this feature assumes the blueprints are for cold dimensions.
